File: problems/problems_1092/solution.py
import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.shortestCommonSupersequence(*test_input)

    def shortestCommonSupersequence(self, str1: str, str2: str) -> str:
        m, n = len(str1), len(str2)
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(m):
            for j in range(n):
                if str1[i] == str2[j]:
                    dp[i + 1][j + 1] = dp[i][j] + 1
                else:
                    dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j])
        i, j = m - 1, n - 1
        ans = []
        while i >= 0 or j >= 0:
            if i < 0:
                ans.append(str2[j])
                j -= 1
                continue
            if j < 0:
                ans.append(str1[i])
                i -= 1
                continue
            if str1[i] == str2[j]:
                ans.append(str1[i])
                i -= 1
                j -= 1
                continue
            if dp[i + 1][j] <= dp[i][j + 1]:
                ans.append(str1[i])
                i -= 1
            else:
                ans.append(str2[j])
                j -= 1
        return ''.join(reversed(ans))

File: problems/problems_1092/Solution.java
package problems.problems_1092;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    public String shortestCommonSupersequence(String str1, String str2) {
        int m = str1.length(), n = str2.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (str1.charAt(i) == str2.charAt(j)) {
                    dp[i + 1][j + 1] = dp[i][j] + 1;
                } else {
                    dp[i + 1][j + 1] = Math.max(dp[i + 1][j], dp[i][j + 1]);
                }
            }
        }
        StringBuilder result = new StringBuilder();
        for (int i = m - 1, j = n - 1; i >= 0 || j >= 0;) {
            if (i < 0) {
                result.append(str2.charAt(j--));
                continue;
            }
            if (j < 0) {
                result.append(str1.charAt(i--));
                continue;
            }
            if (str1.charAt(i) == str2.charAt(j)) {
                result.append(str1.charAt(i));
                --i;
                --j;
            } else if (dp[i + 1][j] <= dp[i][j + 1]) {
                result.append(str1.charAt(i--));
            } else {
                result.append(str2.charAt(j--));
            }
        }
        return result.reverse().toString();
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        String str1 = jsonStringToString(inputJsonValues[0]);
		String str2 = jsonStringToString(inputJsonValues[1]);
        return JSON.toJSON(shortestCommonSupersequence(str1, str2));
    }
}
